File: packages/state/src/position.ts
/**
 * Represents a position in the document (line and column)
 */
export interface Position {
  /** Line number (0-indexed) */
  line: number;
  /** Column number (0-indexed) */
  column: number;
}

/**
 * Create a new position
 */
export function createPosition(line: number, column: number): Position {
  return { line, column };
}

/**
 * Compare two positions
 * Returns negative if a < b, zero if equal, positive if a > b
 */
export function comparePositions(a: Position, b: Position): number {
  if (a.line !== b.line) {
    return a.line - b.line;
  }
  return a.column - b.column;
}

/**
 * Check if two positions are equal
 */
export function positionsEqual(a: Position, b: Position): boolean {
  return a.line === b.line && a.column === b.column;
}

/**
 * Represents a range in the document
 */
export interface Range {
  /** Start position (inclusive) */
  start: Position;
  /** End position (exclusive) */
  end: Position;
}

/**
 * Create a new range
 */
export function createRange(start: Position, end: Position): Range {
  // Ensure start comes before end
  if (comparePositions(start, end) > 0) {
    return { start: end, end: start };
  }
  return { start, end };
}

/**
 * Check if a range is empty (start equals end)
 */
export function isRangeEmpty(range: Range): boolean {
  return positionsEqual(range.start, range.end);
}

/**
 * Check if a range contains a position
 */
export function rangeContains(range: Range, position: Position): boolean {
  return (
    comparePositions(position, range.start) >= 0 &&
    comparePositions(position, range.end) < 0
  );
}

/**
 * Check if two ranges overlap
 */
export function rangesOverlap(a: Range, b: Range): boolean {
  return (
    comparePositions(a.start, b.end) < 0 && comparePositions(b.start, a.end) < 0
  );
}

File: packages/wasm-core/src/position.rs
use serde::{Deserialize, Serialize};

/// Represents a position in the document (line and column)
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct Position {
    /// Line number (0-indexed)
    pub line: usize,
    /// Column number (0-indexed)
    pub column: usize,
}

impl Position {
    pub fn new(line: usize, column: usize) -> Self {
        Position { line, column }
    }

    pub fn zero() -> Self {
        Position { line: 0, column: 0 }
    }
}

impl Default for Position {
    fn default() -> Self {
        Position::zero()
    }
}

/// Represents a range in the document
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct Range {
    /// Start position (inclusive)
    pub start: Position,
    /// End position (exclusive)
    pub end: Position,
}

impl Range {
    pub fn new(start: Position, end: Position) -> Self {
        Range { start, end }
    }

    pub fn from_positions(start_line: usize, start_column: usize, end_line: usize, end_column: usize) -> Self {
        Range {
            start: Position::new(start_line, start_column),
            end: Position::new(end_line, end_column),
        }
    }

    /// Check if the range is empty (start equals end)
    pub fn is_empty(&self) -> bool {
        self.start == self.end
    }

    /// Check if this range contains a position
    pub fn contains(&self, position: Position) -> bool {
        if position.line < self.start.line || position.line > self.end.line {
            return false;
        }
        if position.line == self.start.line && position.column < self.start.column {
            return false;
        }
        if position.line == self.end.line && position.column >= self.end.column {
            return false;
        }
        true
    }
}
